### PR TITLE
Feature/custom build urls

### PR DESCRIPTION
**Pull request by mariangemarcano**
Configure custom urls to show in build status when the build is finished.
the configuration is done master.cfg / BuilderConfig this is an example

```
BuilderConfig(name="my builder",
project="Example Project",
slavenames=["build-slave-04"],
startSlavenames=["build-slave-01"],
factory=factory1,
buildUrls={'Open Tests Tool':
"http://mytool.com/Build/View?"
"configurationUniqueName={builderName}&"
"buildNumber={buildNumber}"}
))
```

### CODE REVIEW

File: master/buildbot/config.py [hunk 1]
```diff
@@ -816,6 +816,34 @@ def __init__(self, name=None, slavename=None, startSlavenames=None, slavenames=N
 
         self.description = description
 
+        self.customBuildUrls = customBuildUrls or {}
```
Comment (BarryUnity) [inline master/buildbot/config.py:819]: What's the advantage of this over making customBuildUrls default to {}?

File: master/buildbot/config.py [hunk 2]
```diff
@@ -816,6 +816,34 @@ def __init__(self, name=None, slavename=None, startSlavenames=None, slavenames=N
...
+        self._validateCustomBuildUrls()
+
+    def _validateCustomBuildUrls(self):
+        """
+        Validates that customBuildUrls  contains strings in the format {'name': 'url'}
```
Comment (BarryUnity) [inline master/buildbot/config.py]: Should be: Validates that customBuildUrls is a dictionary containing only strings and in the format {'name': 'url'} (same goes for the error message)

File: master/buildbot/config.py (hunk 1 above)
Comment (campos-ddc) [reply master/buildbot/config.py:819]: Either make the param `None` (has the benefit of avoiding errors with code altering default values in runtime) or remove this line

File: master/buildbot/config.py [hunk 3]
```diff
@@ -816,6 +816,34 @@ def __init__(self, name=None, slavename=None, startSlavenames=None, slavenames=N
...
+        self._validateCustomBuildUrls()
+
+    def _validateCustomBuildUrls(self):
+        """
+        Validates that customBuildUrls is a dictionary containing only strings and in the format {'name': 'url'}
+        """
+        checkDictionaryContainsOnlyString = all(isinstance(key, str) and isinstance(value, str)
+                                                for key, value in self.customBuildUrls.iteritems())
+
+        if not isinstance(self.customBuildUrls, dict) or (
```
Comment (campos-ddc) [inline master/buildbot/config.py:830]: if `self.customBuildUrls` is not a dict you would already get an error on the line above.

File: master/buildbot/process/botmaster.py [hunk 4]
```diff
@@ -146,6 +146,18 @@ def getBuildernames(self):
...
+        """
+        Gets the BuilderConfiguration by it's name
+        :param name: builder name
+        :return: BuilderConfig
+        :rtype: buildbot.config.BuilderConfig
+        """
+        builderConfig = None
+        if name in self.builders:
+            builderConfig=self.builders[name].config
+        return builderConfig
```
Comment (campos-ddc) [inline master/buildbot/process/botmaster.py]:
```
try:
    return self.builders[name].config
except KeyError:
    return None
```

Also, is `None` an acceptable result? Or should this fail instead?

File: master/buildbot/config.py (hunk 1 above)
Comment (BarryUnity) [reply master/buildbot/config.py:819]: What errors does it avoid?

Comment (mariangemarcano) [reply master/buildbot/config.py:819]: @campos-ddc we need the line to initialize, if the configuration passes None we wont get a failure I can remove it if we prefer a config failure

Comment (campos-ddc) [reply master/buildbot/config.py:819]: @mariangemarcano that is fine, but make the default parameter `None` instead of `{}`

**Review by campos-ddc — APPROVED** [review]

**Pull request merged**